optimade.rst

```diff
@@ -1576,10 +1576,14 @@ Other usual comparison operators ('<', '>', '<=', '>=') MUST NOT be supported.
 It should be noted that boolean values are supposed to be used only in direct comparisons with properties, but not compound comparisons.
 For example, :filter:`(nsites = 3 AND nelements = 3) = FALSE` is not supported.
 
+Boolean property :filter-fragment:`property` MAY be compared with :filter-fragment:`TRUE` by omitting the :filter-fragment:`= TRUE` altogether: :filter:`property`.
```

Comment:
I can't spot anywhere in the spec where we say anything similar for mismatched queries (e.g., querying a list with string operations). Being able to return sensible error messages for this is slightly tricky as the implementation needs to have a type map over all OPTIMADE fields (and its custom fields) - I imagine some are simply attempting the query and catching any errors, which could be dangerous... In the case of the optimade-python-tools reference MongoDB server, you can run any "type" of query on any field, and it can be undefined how the database actually responds (whether the filter returns no entries or all entries) but there will never be an error.

So maybe this is a broader issue...

Reply:
To give a concrete example, due to the way Elasticsearch and MongoDB handle array fields, both NOMAD and odbx can execute the query `elements STARTS WITH "Li"` (coincidentally what you were asking for in the substring search within lists issue).

Should this be forbidden?

Reply:
> To give a concrete example, due to the way Elasticsearch and MongoDB handle array fields, both NOMAD and odbx can execute the query `elements STARTS WITH "Li"` (coincidentally what you were asking for in the substring search within lists issue).

I just wonder what is the semantics of this query – do I get structures that have "Li" as a first element of the array (if yes, what is the ordering of the array), or do I get entries that *have* elements that start with "Li..."? In the latter case, shouldn't it be a "HAS" query?

Reply:
> Should we explicitly say that `property` comparison MUST NOT be supported (HTTP 400 or 501) for non-boolean properties?

It seems that being permissive as possible, provided you can have a sensible and unambiguous result, is a good designe choice in this case.

This brings us back to the policy "truth is non-zero", which would accommodate implementations suggested by @ml-evs . However, I remember during the Jitsy call that @rartino was strongly against such interpretation. When wold be his thoughts on the @ml-evs argument?

Reply:
@ml-evs is right that the specification does not explicitly forbid using list operators on non-list properties, and the same seems to be true vice-versa, non-list operators are not forbidden for lists.

In "Type handling and conversions in comparisons" there are the following sentences:
> Different types of values MUST be reported as `501 Not Implemented` errors, meaning that type conversion is not implemented in the specification.

IMO, this answers [my initial question on this thread](https://github.com/merkys/OPTIMADE/pull/2#discussion_r765873512), but not sure whether this applies for lists vs. non-lists.

Reply:
> > To give a concrete example, due to the way Elasticsearch and MongoDB handle array fields, both NOMAD and odbx can execute the query `elements STARTS WITH "Li"` (coincidentally what you were asking for in the substring search within lists issue).
> 
> I just wonder what is the semantics of this query – do I get structures that have "Li" as a first element of the array (if yes, what is the ordering of the array), or do I get entries that _have_ elements that start with "Li..."? In the latter case, shouldn't it be a "HAS" query?

For MongoDB, it will apply the equivalent string query to each entry in the elements list (so e.g. `elements STARTS WITH "L"` would also work). I agree that it should be a HAS query - the question is whether I need to add code to our implementation to handle the behaviour of this syntactically correct query depending on the `type` of the field being queried, e.g. returning 400 if the query is applied to a list like `elements`. I hadn't actually realised that this query "worked" in our implementation until I tried it today.

> This brings us back to the policy "truth is non-zero", which would accommodate implementations suggested by @ml-evs . However, I remember during the Jitsy call that @rartino was strongly against such interpretation. When wold be his thoughts on the @ml-evs argument?

I think I was in agreement with @rartino on this (which I guess is the inverse case of the above, where a query to a bool field uses non-bool semantics).



> @ml-evs is right that the specification does not explicitly forbid using list operators on non-list properties, and the same seems to be true vice-versa, non-list operators are not forbidden for lists.
> 
> In "Type handling and conversions in comparisons" there are the following sentences:
> 
> > Different types of values MUST be reported as `501 Not Implemented` errors, meaning that type conversion is not implemented in the specification.
> 
> IMO, this answers [my initial question on this thread](https://github.com/merkys/OPTIMADE/pull/2#discussion_r765873512), but not sure whether this applies for lists vs. non-lists.

Yes, I would suggest for this PR (and the overall boolean one) that we do not mandate 400/501 responses specifically for bools, but it might be the case that our reference implementation needs to be fixed to obey these rules. That type handling section is a bit confusing to me really, especially when it comes to lists as you point out.

Reply:
OK, so I will leave Section "Type handling and conversions in comparisons" to cover the comparison of incompatible non-list data types. For list vs. non-list comparison, a separate issue against the specification would be the most appropriate.